### PR TITLE
Fixes issues with From in coqdep (#11631 and #14539) + new unifying semantics of coqdep and coqc

### LINKED ISSUE #11631: coqdep spuriously warns about multiple files with the same name
#### Description of the problem

Given a `B.v` with `From dir.a Require Import A.`, and the directory structure:
```
dir/
    a/
        A.v
    b/
        A.v
    B.v
```
Running `coqdep -R . dir B.v` prints:
```
B.vo B.glob B.v.beautified B.required_vo: B.v a/A.vo
B.vio: B.v a/A.vio
*** Warning: in file B.v, 
    required library A matches several files in path
    (found A.v in a and b; used the latter)
```
So it seems to choose the correct `A.v` despite the warning. The warning goes away if the line is changed to `Require Import dir.a.A.` or if `-Q` is used instead of `-R`.

#### Coq Version

master and at least back to 8.10.

### LINKED ISSUE #14539: coqdep is ignoring `From X` in imports
#### Description of the problem

`coqdep` seems to not take into account the `From` prefix in imports leading to warnings like this:

```
*** Warning: in file theories/WildCat/Adjoint.v, 
    required library Core matches several files in path
    (found Core.v in theories/Metatheory, theories/WildCat, theories/Pointed, theories/Truncations, theories/Spaces/Int, theories/Spaces/Pos, theories/Spaces/No, theories/Categories/PseudonaturalTransformation, theories/Categories/NaturalTransformation, theories/Categories/NaturalTransformation/Composition, theories/Categories/CategoryOfSections, theories/Categories/SetCategory, theories/Categories/Limits, theories/Categories/InitialTerminalCategory, theories/Categories/Adjoint, theories/Categories/Adjoint/UniversalMorphisms, theories/Categories/Adjoint/Composition, theories/Categories/Adjoint/Functorial, theories/Categories/GroupoidCategory, theories/Categories/Structure, theories/Categories/Comma, theories/Categories/Grothendieck/ToSet, theories/Categories/Pseudofunctor, theories/Categories/Category, theories/Categories/Category/Sigma, theories/Categories/KanExtensions, theories/Categories/LaxComma, theories/Categories/Functor/Prod, theories/Categories/Functor, theories/Categories/Functor/Composition, theories/Categories/Functor/Composition/Functorial, theories/Categories/Functor/Pointwise, theories/Categories/Profunctor, theories/Categories/Cat and theories/Categories/FunctorCategory; used the latter)
```
In this case there is an import such as `From WildCat Import Core`. Coq doesn't complain about this but coqdep seems to find every other file named Core not in the WildCat folder.

#### Coq Version
8.15+alpha

### LINKED ISSUE #15053: Merge the doc of `Require` and `From Require`.
Yet another valuable thing to do I think would be to merge the doc of `Require` and `From Require`.

_Originally posted by @herbelin in https://github.com/coq/coq/issues/14718#issuecomment-947864666_

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** fix / enhancements

This PR first addresses the following:
- spurious clash warnings with `From foo Require bar`: this is a hot fix while waiting for @ejgallgo's loadpath cleanup  (#11631, #14539)

It later extended to what is described in the [comment](https://github.com/coq/coq/pull/14718#issuecomment-947862065) below.

Fixes #11631
Fixes #14539
Fixes #15053 

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.

----
Previously, it addressed anomalies or internal errors with non-existent directories and paths but this moved to proper PR. So the following is not anymore in the PR:

<details>
<summary>List of changes initially in this PR but later moved to another PR</summary>

  - `ocamllib`: changing `Uncaught exception: Sys_error("foo: No such file or directory")` into `*** Warning: foo: No such file or directory` on non-existent include (should it be an error or a warning?)
  
     → moved to #14824
  - `coqdep` on non-existent file, after #14024 which fixed a silent ignoring: from `File "foo", line 1: No such file or directory` to `Error: foo: No such file or directory` from the command line, or `*** Warning: In "_FooProject": bar: No such file or directory` if indirectly through a `_CoqProject`

     → moved to #14827

   - `coqdep` on non-existent directory: from a "please report" anomaly to `*** Warning: cannot open foo` (should it be an error?)
     
     → moved to #14826

   - `coqdep`: when a file exists but not with the expected `From root`, clarify the message that it does not exist only wrt this root

     → moved to #14829

  - `coq_makefile` on non-existent project file: changing `Fatal error: exception Sys_error("foo: No such file or directory")` into `Error: foo: No such file or directory`
   
  → moved to #14828

  - `coqdep` failing nicely and with better message where it used to fail with an anomaly of the form `Unknown option /home/coq/foo` when reading a Coq project with dirname `/home/coq` and some uninterpretable string `foo` is present in the file [Added 8/8/21]
  
  → moved to #14825

  - `coq_makefile` accepts option `-impredicative-set` when on the command line (or should it be `-arg -impredicative-set` even on the command line??) [Added 8/8/21]
  
  → abandoned

  - `coqdep` fails nicely rather than with an anomaly on parsing error of a project file [Added 8/8/21]
   
  → moved to #14828
</details>

### CODE REVIEW

**Comment by ppedrot** [comment]
> this is a hot fix

@herbelin last time I had a look the code was completely nonsensical. I don't think your code restores soundness at all, it's just hiding the nonsense under the rug and is probably going to break some developments in fancy ways. I'd recommend kicking @ejgallego 's ass for the PR development to go faster rather than piling hacks over hacks in this very broken part of the code. Worst case that qualifies as a reasonable hotfix is to copy Loadpath into coqdep and remove it when the factoring actually happens.

**Comment by herbelin** [comment]
Interestingly, the PR reveals that in the presence of overlapping between `-R`/`-Q` options, there are different kinds of overlapping. Let's consider the following layout:
```
├── B1
│   └── C
│   |   └── E.v
│   └── D
│       └── E.v
├── B2
    └── E.v
```
In `coqdep -R B1 B -R B2 B`, there are ambiguities of different kinds: the resolution of `E` between `B.C.E` and `B.D.E` is determined by the order between `C` and `D` in the file system (knowing that the underlying hash table fortunately preserves the order of entry in its buckets), but the discrimination with `B.E` depends only in the order of arguments on the command line. So, we may like to warn only in the first case and not in the second case.

This is however not what `coqdep` does: it always warns and does not take into account that the order in the command line gives a criterion to decide even though both it and `coqc` actually use this criterion to take a decision.

**Comment by herbelin** [comment]
> the code was completely nonsensical. I don't think your code restores soundness at all...

What kind of "nonsensical" do you have in mind? Barring corner cases, it seems these days to reasonably work? (I'm not saying that the code should not be shared with `coqc`, on the contrary, but it is relatively better now than say 10 years ago, at a time where `coqc` and `coqdep` quickly disagreed).

Moreover, the warnings I added (as a compensation of the ones I removed) allow to reveal that e.g. hierarchy-builder was talking some risks in having two files named the same in the same file tree, w/o probably having realized they were doing it.

File: lib/coqProject_file.ml [hunk 1]
```diff
@@ -116,12 +116,11 @@ and parse_args buf accu s = match Stream.next s with
 | exception Stream.Failure -> accu
 
 let parse f =
-  let c = open_in f in
+  let c = try open_in f with Sys_error msg -> Format.eprintf "Error: @[%s@].@\n%!" msg; exit 1 in
```
Comment (SkySkimmer) [inline lib/coqProject_file.ml:119]: Do we really want this to be unrecoverable? Seems like it should just raise a Parsing_error

Comment (herbelin) [reply lib/coqProject_file.ml:119]: You mean returning a parsing error if the project file to open cannot be opened? This looks confusing to me.

Comment (SkySkimmer) [reply lib/coqProject_file.ml:119]: Better than exit imo

Comment (Alizter) [reply lib/coqProject_file.ml:119]: Perhaps we need an `UnableToOpenProjectFile` exception that reraises this exception. That way it is clear that the project file was unable to be opened and the reason why. I agree that exiting is probably not the best thing to do.

File: tools/coqdep_common.ml [hunk 2]
```diff
@@ -127,60 +129,74 @@ let get_prefix p l =
...
-      | None -> ()
-      | Some _ -> raise (M.Found binding)
+      | None -> acc
+      | Some _ -> get binding :: acc
+    else
+      acc
   in
-  try Hashtbl.iter iter table; raise Not_found
-  with M.Found s -> s
+  (* We reverse to give priority for compatibility. There is an issue
```
Comment (Alizter) [inline tools/coqdep_common.ml:145]: Can you expand on the compatibility requirements. Is anyone relying on this behaviour?

File: lib/coqProject_file.ml (hunk 1 above)
Comment (SkySkimmer) [reply lib/coqProject_file.ml:119]: I don't see what's confusing about `Parsing_error "Could not open project file foo"` tbh
alternatively we don't need to catch the sys_error

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
Do not `exit 1`

File: lib/coqProject_file.ml (hunk 1 above)
Comment (herbelin) [reply lib/coqProject_file.ml:119]: As `coqdep` is currently done, it reports uncaught errors as anomalies, so we need to do something. I'd be tempted to adopt the `UnableToOpenProjectFile` proposal.

File: tools/coqdep_common.ml (hunk 2 above)
Comment (herbelin) [reply tools/coqdep_common.ml:145]: hierarchy-builder was relying on this order.

**Comment by herbelin** [comment]
> Do not `exit 1`

I added an exception `UnableToOpenProjectFile` but we still need to eventually do something when the exception is raised. I don't know what else to do than reporting a failure and exiting. Tell me if you have something else in mind.

**Review by SkySkimmer — APPROVED** [review]

**Comment by ppedrot** [comment]
It looks like this PR is turning into a hodgepodge of random fixes that are loosely related and not even for the same binaries (`coq_makefile` vs. `coqdep`). Couldn't we get instead smaller PRs that would be merged faster and more easily?

Also, my main remark still stands: coqdep is just as broken as before and no amount of haphazard fiddling with the set of clashes is going to fix it. You need to completely rework the thing for it to be sound.

**Comment by herbelin** [comment]
Yes, it is a hodgepodge of not random but indeed heteroclitous fixes while observing various problems of coqdep, ocamllibdep and coq_makefile.

Sure, it can be split. Will do.

About reworking the whole thing, it will not be by me, but I will not cry for the current coqdep if someone finds the courage to rewrite it.

**Comment by herbelin** [comment]
As another divergence between `coqdep` and `coqc`, in:
```
├── B
│   └── D.v  (* 1 *)
│   └── C
│       └── D.v (* 2 *)
```
`coqc` treats the files of `B` before the ones of `B/C` leading to give priority to instance 1 of `D.v` while `coqdep` treats files and subdirectories in a same pass, so depending on the internal order of the contents of B, treating 1 first if `D.v` comes before `C` in the order given by `Sys.readdir` for directory `B` and treating 2 first otherwise.

**Comment by herbelin** [comment]
> I'd recommend kicking @ejgallego 's ass for the PR development to go faster rather than piling hacks over hacks in this very broken part of the code. Worst case that qualifies as a reasonable hotfix is to copy Loadpath into coqdep and remove it when the factoring actually happens.

@ejgallego, @ppedrot: how would you rewrite `coqdep`? In addition to preserving order, I believe efficiency is important, no? Would `dune` comes with support for what we need? Otherwise, would you typically have in mind a tree-like structure with access first from the top following the qualifiers of the `From` field then going from the bottom following the qualifiers of the module to find, as e.g.:
```
type entry = file_name * logical_name * original_include_command
type ('a,'b) tree = Leaf of 'a | Node of 'b
type rev_dict = (entry list, rev_dict) tree StringMap.t
type dict = (rev_dict, dict) tree StringMap.t
```
In any case, this would not prevent being cautious about the order files are added to the structure, deciding when to override and when to report a clash.

I'm afrait including `Loadpath` would be costly...

**Comment by herbelin** [comment]
Another parameter to take into account about ordering:
```
├── B
│   └── E.v (* 1 *)
│   └── C
│       └── E.v (* 2 *)
│       └── D
│           └── E.v (* 3 *)
```
do we go upwards or downwards? `coqc` goes in a mixed order 1, 3, 2 and, probably, we should keep this order for compatibility. Or we can also adopt the more intuitive order 1, 2, 3? The reason for the mixed order is that the top directory is treated separately. So with more nestings, it would be 1, n+1, n, ..., 2.

**Comment by ejgallego** [comment]
@herbelin indeed we likely need to tweak the current semantics a bit, but we need to put all the loadpath code in one place first.

**Comment by ejgallego** [comment]
@herbelin thanks for all the careful and detailed analysis here, it is indeed super useful.

**Comment by herbelin** [comment]
> @herbelin indeed we likely need to tweak the current semantics a bit, but we need to put all the loadpath code in one place first.

One thing about sharing the loadpath code is that `coqc` works in a lazy way in the sense that it builds first the loadpath structure without looking at the files, then, for each `Require`, looks (dynamically) in the different possible directories. I don't know how much this is really useful, but, for instance, you can compile files after having started the editor and have the files taken into account only dynamically.

On the other side, mostly for efficiency reasons, `coqdep` does not build a loadpath structure but instead directly build a map binding qualified names (or rather pairs of from qualifier and require qualifiers) to filenames.

To have all the loadpath code in one place, do you lean towards relying on the coqc model or on the coqdep model?

**Comment by herbelin** [comment]
I gave in to and implemented a direct map from pairs `(from,suffix)` with a hash-table. Execution times seem equivalent (that is, most of the time is spent in building the `-R coqlibpath Coq` map!).

This allows to be independent of the order of `Hashtbl.fold` and to fix the [second part](https://github.com/coq/coq/issues/14539#issuecomment-895501948) of #14539. This also somehow answers a comment of @ppedrot in 31c754273 ("We should rather use some trie-based data structure, but I'll leave this for a future commit."). [Note that this does not prevent at all a cleanup of loadpath management in general, on the contrary. The knowledge we acquired in the process will be useful for a cleanup.]

I added more tests showing how ambiguities are resolved when a file occurs several times in the same `-Q`/`-R` subspace showing the dependency on the file system (that is on the order given by `Sys.readdir`). As a result, the Windows test and my own setup show a different resolution of the ambiguity than the MacOS and base tests! I then don't see what alternatives there are than forbidding such `Require`s whose resolution rely on the filesystem order??

**Comment by ppedrot** [comment]
>  I then don't see what alternatives there are than forbidding such Requires whose resolution rely on the filesystem order??

Why not normalize the output of `readdir`? The OCaml doc explicitly states that
> There is no guarantee that the name strings in the resulting array will appear in any specific order; they are not, in particular, guaranteed to appear in alphabetical order.

**Comment by herbelin** [comment]
> Why not normalize the output of readdir?

You mean using byte order? Yes, that's a possibility also.

There might be utf8 issues though when working across different platforms. A same unicode file can be encoded in different utf8 ways depending on file systems (e.g. on MacOS X, accented letters are encoded in filenames with combining accents).

**Comment by ppedrot** [comment]
> You mean using byte order?

Not necessarily with byte order. As you mention, we could also perform Unicode normalization. That said, it's not clear that we wouldn't already have other similar issues with cross-platform binary vo files in coqc, since we handle identifiers as mere binary strings.

**Comment by herbelin** [comment]
> > You mean using byte order?
> 
> Not necessarily with byte order. As you mention, we could also perform Unicode normalization.

You mean Unicode normalization + `Stdlib.compare`? Or collation? The first is maybe easier according to [uucp](https://erratique.ch/software/uucp/doc/unicode.html#eqcmpnorm)'s page.

> That said, it's not clear that we wouldn't already have other similar issues with cross-platform binary vo files in coqc, since we handle identifiers as mere binary strings.

I believe we embed strings the form they were entered, even the few filenames stored in locations.

As for filenames, (now old) HFS+ for instance [unicode-normalizes](https://developer.apple.com/library/archive/technotes/tn/tn1150.html#UnicodeSubtleties) filenames before doing comparison. New APFS does not normalize itself but MacOS libraries apparently normalize before communicating with APFS, according to this [page](https://eclecticlight.co/2021/05/08/explainer-unicode-normalization-and-apfs/).

Incidentally, should we normalize the kernel names, or at least normalize the qualids before asking the nametab?

**Comment by ejgallego** [comment]
Sorry folks I am not following the discussion closely, but something I am positive based on experience with large builds is that all directory traversal has to be lazy if possible.

**Comment by herbelin** [comment]
> Sorry folks I am not following the discussion closely, but something I am positive based on experience with large builds is that all directory traversal has to be lazy if possible.

As a matter of fact, most of the time spent in a call to `coqdep` is to reconstruct the map for the standard library. It should indeed be done lazily or maybe even cached somewhere.

**Comment by herbelin** [comment]
As far as I can judge, the error in `perennial` was "real": `coqdep` and `coqc` disagreed on how to manage the loadpath but since, thanks to `coqdep`, only one of the two conflicting files was compiled, the vo file that `coqc` would have taken by default was absent and the file mentioned by `coqdep` was taken.

Told in another way, `coqdep` may impact the choices made by `coqc` by imposing an order of compilation to `coqc`.

So, it is not clear that `coqdep` should necessarily comply with `coqc`. It can be `coqc` which align on `coqdep`.

The last commit I added makes that `coqc` follows `coqdep` on the order in which subdirectories are considered (a `List.rev` is added).

(The problem of `Sys.readdir`'s order being platform or file-system dependent remains though.)

**Review by Alizter — COMMENTED** [review]
Here are some misc suggestions before I review the ml.

File: test-suite/misc/deps-order-from.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+$coqc -R misc/deps/test-from T misc/deps/test-from/B/C.v 2>&1
+$coqc -R misc/deps/test-from T misc/deps/test-from/D.v 2>&1
+$coqc -R misc/deps/test-from T misc/deps/test-from/E.v 2>&1
+S=$?
+if [ $R = 0 ] && [ $S = 0 ]; then
+    printf "coqdep and coqc agree\n"
+    exit 0
+else
+    printf "coqdep and coqc disagree\n"
+    exit 1
```
[the remark this answers is not in the document]
Comment (herbelin) [reply test-suite/misc/deps-order-from.sh:21]: You mean adding two periods right?

[BTW, note that the message does not exactly capture agreement/disagreement but also failure of one and not failure of the other or such kind of things. I did not take time to fully investigate what the message captured and I left the current approximation.]

File: test-suite/misc/deps-order-from.sh [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+rm -f misc/deps/test-from/A/C.vo misc/deps/test-from/B/C.vo misc/deps/test-from/D.vo misc/deps/test-from/E.vo
+tmpoutput=$(mktemp /tmp/coqcheck.XXXXXX)
+$coqdep -R misc/deps/test-from T misc/deps/test-from/D.v misc/deps/test-from/E.v 2>&1 > "$tmpoutput"
+diff -u --strip-trailing-cr misc/deps/deps-from.out "$tmpoutput" 2>&1
+R=$?
+times
+$coqc -R misc/deps/test-from T misc/deps/test-from/A/C.v 2>&1
+$coqc -R misc/deps/test-from T misc/deps/test-from/B/C.v 2>&1
+$coqc -R misc/deps/test-from T misc/deps/test-from/D.v 2>&1
+$coqc -R misc/deps/test-from T misc/deps/test-from/E.v 2>&1
```
Comment (SkySkimmer) [inline test-suite/misc/deps-order-from.sh:14]: Why all the `2>&1`?

**Review by Alizter — COMMENTED** [review]
Just this comment about exceptions for now. Will review more tomorrow.

File: tools/coqdep_common.ml [hunk 5]
```diff
@@ -115,45 +109,22 @@ let mkknown () =
...
-    | None -> ()
-    | Some rem ->
-      match get_prefix (List.rev s) (List.rev rem) with
-      | None -> ()
-      | Some _ -> raise (M.Found binding)
-  in
-  try Hashtbl.iter iter table; raise Not_found
-  with M.Found s -> s
+let search_table table ?(from=[]) s =
+  Hashtbl.find table (from, s)
```
[the remark this answers is not in the document]
Comment (herbelin) [reply tools/coqdep_common.ml:120]: I kept it as it was, but we could use `find_opt` as well. I can do it if you insist.

File: test-suite/misc/deps-order-from.sh (hunk 4 above)
Comment (herbelin) [reply test-suite/misc/deps-order-from.sh:14]: If there are warnings, we may want to know about them, no?

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
I will merge tomorrow if @ppedrot is happy with it

**Comment by herbelin** [comment]
Note that for `From`, it uses now a direct access in a hash table while before it was looking in a list. So, I suspect that performance can only be better in general.

Note also that performance-wise, a lot of the time is spent to reconstruct again and again the file map for the stdlib, so a cache would help here if someone knows how to build one (for coqdep on a single file, rebuilding the stdlib map is close to 100% of the time).

**Comment by herbelin** [comment]
> Bench indicates no issue

Aren't times around 0 on the average in general? (Anyway, I don't see a reason why coquelicot would be slower because of coqdep.)

**Comment by herbelin** [comment]
I finally implemented a variant of what was discussed at [Zulip](https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs.20.26.20plugin.20devs/topic/Directory.20handling.20code.20in.20coqdep):

- If some files exactly match the required logical name, they are selected. If there is a `From` clause, this uniquely determines a file. If there is no `From` clause, the file corresponding to the most recently given `-R` or `-Q` is selected.
- If no files exactly match the required logical name, the ones corresponding to the most recently given `-R` or `-Q` (or to the `From` prefix if any) are selected. If there are several of them (in case of `-R`), the choice is system-dependent. [It could be possible to raise an error with a little extra effort in `coqc`, but I did not make this effort yet.]

This change of semantics seems rather natural and it provides a way to non-ambiguously refer to a given file, thus answering to this [Coq Club discussion](https://sympa.inria.fr/sympa/arc/coq-club/2020-09/msg00043.html). It is also compatible with the CI.

**Review by jfehrle — CHANGES_REQUESTED** [review]
Hope you will improve the wording in `modules.rst`.

IMHO much of that section (not just your changes) deserves to be rewritten.  Also, I would consider other improvements to the described functionality--though not in this PR.

**Comment by Alizter** [comment]
@herbelin Can you have a look at Jim's review? I want to merge this soon, but we should finish polishing up the doc.

**Comment by herbelin** [comment]
> > I'm now also leaning in the direction to fail if a `Require` is ambiguous.
> 
> A warning (and ignoring one file) would be more user-friendly.

You mean as soon as `coqdep`?

> The conflict may be unimportant. The message should be specific about the conflicting files.

Yes, and the code for reporting the conflict is already done (with a warning).

> If there are several conflicts, they should all be reported. I expect `coqc` exits after reporting the first error, which would be less friendly.

This is one of the reason which pushes me to prefer a warning in `coqdep` (but the warning should not be the current warning, it should be something about "ambiguous path, it would fail at the time of compiling" or something like that)

> > The order in which subdirectories of `user-contrib` are added to the loadpath is dependent on the file system.
> 
> How about sorting the list of filenames in each directory before loading them? That would make it not system dependent. Sorting unicode correctly (from the user perspective) is locale-dependent, so that's not a good idea. Sorting on Unicode in some fixed locale or just doing a byte comparision of the Unicode would be sufficient to make it not system dependent.

Sorting with Unicode was discussed [above](https://github.com/coq/coq/pull/14718#issuecomment-913134938). One would have to rely on a unicode library, and it'd be a bit of work to implement, but why not.

> > if there are `Foo1` and `Foo2` in `user-contrib`, and both contains a file `Bar`, `Require Bar` will be system dependent. One answer would be to force to use `From` whenever we require a file from an installed package? Is it acceptable?
> 
> The shorter form is nice for users. Perhaps the code could generate an error only when `Bar` is ambiguous?

I was suggesting a `From`, because it would then be robust to addition of new possibly-newly-conflicting libraries in `user-contrib`.

**Review by jfehrle — COMMENTED** [review]
Getting there.  How about these additional suggestions?

**Comment by SkySkimmer** [comment]
>Option -Q does not require From, if this is what you mean.

I mean for instance I cannot do `Require Ltac2` I need to do `Require Ltac2.Ltac2` or `From Ltac2 Require Ltac2`, so even if someone installs `user-contrib/foo/Ltac2.vo` there is no confusion.

**Comment by jfehrle** [comment]
OTOH, it occurs to me that if the Coq strongly discourages having ambiguous files, then maybe Unicode sorting isn't really necessary.

**Comment by herbelin** [comment]
Another issue with Unicode is that the loadpath is built by looking at directory names as they are in the file system (that is Unicode-normalized if on a file system that normalizes), but the correspondence with the logical path is done without normalization. Either a normalization function should be used (with applicability to all file systems) or the normalization function of the underlying file system (if any) is used to normalize (e.g. by creating a tmp file and reading its normalized name back).

**Comment by herbelin** [comment]
Sorry for yet an iteration on the PR. I modified the following behaviors:
- `coqc`/`coqtop` now fail if a logical path partially matches two files in the same `-R`/`-Q` root; on its side, `coqdep` warns that `Require` will fail (maybe `coqdep` should already fail as it currently produces dependency files which are file system and that's a bit strange)
- `coqc`/`coqtop` do not any longer warn when two exact matches exist; instead, they take the latest; `coqdep` warns that the latest will be used
- also refined `Load`: `Load id` searches in path, but `Load "basename/filename"`, which is a non-sense in terms of logical path, does not

Note that, eventually, a few commits will have to be smashed. I'm also open to split the PR again in several smaller PRs if it helps.

**Comment by herbelin** [comment]
> Any clue what the issue with equations is?

It is an ambiguous `Require`, see mattam82/Coq-Equations#435.

**Review by Alizter — APPROVED** [review]

**Comment by herbelin** [comment]
I squashed, making the different parts of the PR more visible:
- Fix of #11631 (initial report) and #14539
- Misc internal cleaning/renaming in coqdep_common.ml
- Adopting same order for browsing directories in coqc and coqdep
- A mini-fix in an error message introduced in #14827
- Removing an outdated comment in the refman
- An update of coqdep's man page and refman
- The new exact-match priority, order on `-R`/`-Q`, failure on ambiguity rules

**Comment by ejgallego** [comment]
This seems good to go, great effort by @herbelin !

We have many options on how to merge the two codepaths, there is not a canonical strategy IMHO, as of today, I think that just make `coqdep` directly depend on `Loadpath`, even if `Loadpath`still depends on `coq-core.library` is the best way to go.

For that you need a small commit, see: https://github.com/ejgallego/coq/commit/495795f13e2c93f26ddfca8b391b585289a44242

**Review by Zimmi48 — COMMENTED** [review]
I came up to this PR through the issue opened by @Alizter at #15053, only to discover that this merge was already attempted here, so here are my recommendations on completing it (+ some comments on the changelog).

**Comment by herbelin** [comment]
> Caching is IMHO an orthogonal issue, tries could support both [and I was under the impression we already do caching in the `coqc` case, don't we @herbelin] ?

There is (or at least there was at some time) a caching of loaded `vo` files, but looking for an existing `vo` files was only at the time of requiring.

I don't have a strong opinion on the data structures to use, otherwise. Feel free to experiment what you think is good.

**Review by Zimmi48 — COMMENTED** [review]
This updated documentation looks great! (And the change to the semantics also is great!)

FTR, my plan as part of the (still ongoing) restructuring of the reference manual was to get rid of the "Commands" chapter (formerly "Vernacular commands") and sprinkle out the commands where they make most sense. In particular, the `Require` command should move to the "Libraries and filesystem" section (thus closer to where `Import` and `Export` are documented).

Since you are already changing many things in this part of the documentation, it could make sense to also do the move as part of this PR. But that's entirely up to you.

**Review by jfehrle — APPROVED** [review]
LGTM.  A last few minor tweaks.

**Review by Zimmi48 — APPROVED** [review]

**Review by ejgallego — APPROVED** [review]

**Pull request merged**